File: go-version/leetcode/LC1578/min_cost.go
package LC1578

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/11/3 下午11:22
 * @FilePath: min_cost
 * @Description: 贪心 + 栈
 */

func minCost(colors string, neededTime []int) int {
	preColor := colors[0]
	preTime, cost := neededTime[0], 0
	for i := 1; i < len(colors); i++ {
		if preColor == colors[i] {
			cost += min(preTime, neededTime[i])
			preTime = max(preTime, neededTime[i])
		} else {
			preColor = colors[i]
			preTime = neededTime[i]
		}
	}
	return cost
}

File: python-version/LC1578/min_cost.py
from typing import List


class Solution:
    @staticmethod
    def minCost(colors: str, needed_time: List[int]) -> int:
        pre_color = colors[0]
        pre_time, cost = needed_time[0], 0
        for i in range(1, len(colors)):
            if pre_color == colors[i]:
                cost += min(pre_time, needed_time[i])
                pre_time = max(pre_time, needed_time[i])
            else:
                pre_color = colors[i]
                pre_time = needed_time[i]
        return cost
